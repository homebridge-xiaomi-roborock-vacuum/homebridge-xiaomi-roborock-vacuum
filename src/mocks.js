// ============= MIIO MOCKS ================

const miioDevice = {
  miioModel: "test-model",
  matches: jest.fn(),
  destroy: jest.fn(),
  property: jest.fn(),
  on: jest.fn(),
  handle: { api: { parent: { socket: {} } } },
  poll: jest.fn(),
  state: jest.fn().mockReturnValue({}),
};

const miio = {
  device: miioDevice,
  createMock: () => ({
    device: jest.fn().mockImplementation(() => miioDevice),
  }),
};

// ============= HOMEBRIDGE MOCKS ================

const createGetCharacteristicMock = () =>
  jest.fn().mockImplementation(() =>
    Object.assign({}, createChainableServiceMethodsMock(), {
      setProps: jest.fn(),
      on: createGetCharacteristicMock(),
      updateValue: jest.fn(),
    })
  );

const createChainableServiceMethodsMock = () => ({
  getCharacteristic: createGetCharacteristicMock(),
  setCharacteristic: jest
    .fn()
    .mockImplementation(createChainableServiceMethodsMock),
});

const createServiceMock = () =>
  jest
    .fn()
    .mockImplementation((name, type) =>
      Object.assign({ name, type }, createChainableServiceMethodsMock())
    );

const Service = createServiceMock();
Service.Switch = createServiceMock();
Service.FilterMaintenance = createServiceMock();
Service.AccessoryInformation = createServiceMock();
Service.Fan = createServiceMock();
Service.BatteryService = createServiceMock();
Service.OccupancySensor = createServiceMock();

const Characteristic = jest.fn();
Characteristic.Formats = { FLOAT: 1 };
Characteristic.Perms = { READ: 1, NOTIFY: 2 };
Characteristic.BatteryLevel = jest.fn();
Characteristic.StatusLowBattery = jest.fn();
Characteristic.StatusLowBattery.BATTERY_LEVEL_LOW = "BATTERY_LEVEL_LOW";
Characteristic.StatusLowBattery.BATTERY_LEVEL_NORMAL = "BATTERY_LEVEL_NORMAL";
Characteristic.ChargingState = jest.fn();
Characteristic.ChargingState.CHARGING = "CHARGING";
Characteristic.ChargingState.NOT_CHARGING = "NOT_CHARGING";
Characteristic.OccupancyDetected = jest.fn();
Characteristic.On = jest.fn();
Characteristic.RotationSpeed = jest.fn();
Characteristic.FilterChangeIndication = jest.fn();
Characteristic.FilterLifeLevel = jest.fn();
Characteristic.Manufacturer = jest.fn();
Characteristic.Model = jest.fn();
Characteristic.FirmwareRevision = jest.fn();
Characteristic.SerialNumber = jest.fn();

const createHomebridgeMock = () => ({
  registerAccessory: jest.fn(),
  hap: { Characteristic, Service },
});

module.exports = {
  createHomebridgeMock,
  miio,
};
