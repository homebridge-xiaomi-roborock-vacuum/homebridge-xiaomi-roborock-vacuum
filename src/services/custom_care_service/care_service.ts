import { Service } from "hap-nodejs";
import {
  CareFilter,
  CareMainBrush,
  CareSensors,
  CareSideBrush,
} from "./characteristics";

export class Care extends Service {
  public static UUID = "00000111-0000-0000-0000-000000000000";
  constructor(displayName: string, subType?: string) {
    super(displayName, Care.UUID, subType);
    this.addCharacteristic(CareSensors);
    this.addCharacteristic(CareFilter);
    this.addCharacteristic(CareSideBrush);
    this.addCharacteristic(CareMainBrush);
  }
}
