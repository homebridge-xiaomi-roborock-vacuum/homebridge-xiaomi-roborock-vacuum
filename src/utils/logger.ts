import { Logging } from "homebridge";

export interface CustomLoggerConfig {
  /**
   * When `true`, info and debug logs are silenced to avoid convoluted logs.
   */
  silent?: boolean;
}

const noop = () => {};

export interface Logger
  extends Pick<Logging, "debug" | "info" | "warn" | "error"> {
  setModel: (modelName: string) => void;
}

export function getLogger(log: Logging, config: CustomLoggerConfig): Logger {
  let model = "unknown";

  function buildMsg(message: string) {
    return `[Model=${model}] ${message}`;
  }

  return {
    debug: (msg, ...params) =>
      config.silent ? noop() : log.debug(buildMsg(msg), ...params),
    info: (msg, ...params) =>
      config.silent ? noop() : log.info(buildMsg(msg), ...params),
    warn: (msg, ...params) => log.warn(buildMsg(msg), ...params),
    error: (msg, ...params) => log.error(buildMsg(msg), ...params),
    setModel: (modelName) => (model = modelName),
  };
}
