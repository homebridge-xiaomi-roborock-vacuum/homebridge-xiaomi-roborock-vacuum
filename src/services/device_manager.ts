import { HAP } from "homebridge";
import { BehaviorSubject, distinct, exhaustMap, Subject, timer } from "rxjs";
import miio from "../miio";
import { Logger } from "../utils/logger";
import { MiioDevice } from "../utils/miio_types";
import { cleaningStatuses } from "../utils/constants";

export interface DeviceManagerConfig {
  ip?: string;
  token?: string;
}

export interface ErrorChangedEvent {
  id: string;
  description: unknown;
}

export interface StateChangedEvent {
  key: string;
  value: unknown;
}

const GET_STATE_INTERVAL_MS = 30000; // 30s

export class DeviceManager {
  private readonly internalDevice$ = new BehaviorSubject<
    MiioDevice | undefined
  >(undefined);

  private readonly ip: string;
  private readonly token: string;

  private readonly internalErrorChanged$ = new Subject<ErrorChangedEvent>();
  private readonly internalStateChanged$ = new Subject<StateChangedEvent>();
  public readonly errorChanged$ = this.internalErrorChanged$.pipe(distinct());
  public readonly stateChanged$ = this.internalStateChanged$.asObservable();
  public readonly deviceConnected$ = this.internalDevice$.asObservable();

  private connectingPromise: Promise<void> | null = null;
  private connectRetry = setTimeout(() => void 0, 100); // Noop timeout only to initialise the property
  constructor(
    private readonly hap: HAP,
    private readonly log: Logger,
    config: DeviceManagerConfig
  ) {
    if (!config.ip) {
      throw new Error("You must provide an ip address of the vacuum cleaner.");
    }
    this.ip = config.ip;

    if (!config.token) {
      throw new Error("You must provide a token of the vacuum cleaner.");
    }
    this.token = config.token;

    this.connect().catch(() => {
      // Do nothing in the catch because this function already logs the error internally and retries after 2 minutes.
    });
  }

  public get model() {
    return this.internalDevice$.value?.miioModel || "unknown model";
  }

  public get state() {
    return this.property("state") as string;
  }

  public get isCleaning() {
    return cleaningStatuses.includes(this.state);
  }

  public get isPaused() {
    return this.state === "paused";
  }

  public get device() {
    if (!this.internalDevice$.value) {
      throw new Error("Not connected yet");
    }
    return this.internalDevice$.value;
  }

  public property<T>(propertyName: string) {
    return this.device.property<T>(propertyName);
  }

  public async ensureDevice(callingMethod: string) {
    try {
      if (!this.internalDevice$.value) {
        const errMsg = `${callingMethod} | No vacuum cleaner is discovered yet.`;
        this.log.error(errMsg);
        throw new Error(errMsg);
      }

      // checking if the device has an open socket it will fail retrieving it if not
      // https://github.com/aholstenson/miio/blob/master/lib/network.js#L227
      const socket = this.internalDevice$.value.handle.api.parent.socket;
      this.log.debug(
        `DEB ensureDevice | ${this.model} | The socket is still on. Reusing it.`
      );
    } catch (error) {
      const err = error as Error;
      if (
        /destroyed/i.test(err.message) ||
        /No vacuum cleaner is discovered yet/.test(err.message)
      ) {
        this.log.info(
          `INF ensureDevice | ${this.model} | The socket was destroyed or not initialised, initialising the device`
        );
        await this.connect();
      } else {
        this.log.error(err.message, err);
        throw err;
      }
    }
  }

  private async connect() {
    if (this.connectingPromise === null) {
      // if already trying to connect, don't trigger yet another one
      this.connectingPromise = this.initializeDevice().catch((error) => {
        this.log.error(
          `ERR connect | miio.device, next try in 2 minutes | ${error}`
        );
        clearTimeout(this.connectRetry);
        // Using setTimeout instead of holding the promise. This way we'll keep retrying but not holding the other actions
        this.connectRetry = setTimeout(
          () => this.connect().catch(() => {}),
          120000
        );
        throw error;
      });
    }
    try {
      await this.connectingPromise;
      clearTimeout(this.connectRetry);
    } finally {
      this.connectingPromise = null;
    }
  }

  private async initializeDevice() {
    this.log.debug("DEB getDevice | Discovering vacuum cleaner");

    const device = await miio.device({ address: this.ip, token: this.token });

    if (device.matches("type:vaccuum")) {
      this.internalDevice$.next(device);

      this.log.setModel(this.model);

      this.log.info("STA getDevice | Connected to: %s", this.ip);
      this.log.info("STA getDevice | Model: " + this.model);
      this.log.info("STA getDevice | State: " + this.property("state"));
      this.log.info("STA getDevice | FanSpeed: " + this.property("fanSpeed"));
      this.log.info(
        "STA getDevice | BatteryLevel: " + this.property("batteryLevel")
      );

      this.device.on<ErrorChangedEvent>("errorChanged", (error) =>
        this.internalErrorChanged$.next(error)
      );
      this.device.on<StateChangedEvent>("stateChanged", (state) =>
        this.internalStateChanged$.next(state)
      );

      // Refresh the state every 30s so miio maintains a fresh connection (or recovers connection if lost until we fix https://github.com/homebridge-xiaomi-roborock-vacuum/homebridge-xiaomi-roborock-vacuum/issues/81)
      timer(0, GET_STATE_INTERVAL_MS).pipe(exhaustMap(() => this.getState()));
    } else {
      const model = (device || {}).miioModel;
      this.log.error(
        `Device "${model}" is not registered as a vacuum cleaner! If you think it should be, please open an issue at https://github.com/homebridge-xiaomi-roborock-vacuum/homebridge-xiaomi-roborock-vacuum/issues/new and provide this line.`
      );
      this.log.debug(device);
      device.destroy();
    }
  }

  private async getState() {
    try {
      await this.ensureDevice("getState");
      await this.device.poll();
      const state = await this.device.state();
      this.log.debug(
        `DEB getState | ${this.model} | State %j | Props %j`,
        state,
        this.device.properties
      );

      for (const key in state) {
        if (key === "error") {
          this.internalErrorChanged$.next(state[key]);
        } else {
          this.internalStateChanged$.next({ key, value: state[key] });
        }
      }
    } catch (err) {
      this.log.error(`getState | %j`, err);
    }
  }
}
