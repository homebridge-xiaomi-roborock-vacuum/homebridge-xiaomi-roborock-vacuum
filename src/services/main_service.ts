import { Service } from "homebridge";
import { distinct, filter, map, tap } from "rxjs";
import { findSpeedModes } from "../utils/find_speed_modes";
import { CoreContext } from "./types";
import type { RoomsService } from "./rooms_service";
import type { ProductInfo } from "./product_info";
import { PluginServiceClass } from "./plugin_service_class";
import { ensureName } from "../utils/ensure_name";

export interface MainServiceConfig {
  serviceType: "fan" | "switch";
}

export class MainService extends PluginServiceClass {
  public readonly cachedState = new Map<string, unknown>();
  private readonly service: Service;
  constructor(
    coreContext: CoreContext,
    private readonly productInfo: ProductInfo,
    private readonly roomsService: RoomsService,
    private readonly setWaterSpeed: (mode: string) => Promise<void>,
    private readonly changedPause: (isCleaning: boolean) => void
  ) {
    super(coreContext);
    if (this.config.serviceType === "fan") {
      this.service = new this.hap.Service.Fan(this.config.name, "Vacuum");
      this.service
        .getCharacteristic(this.hap.Characteristic.RotationSpeed)
        .onGet(() => this.getSpeed())
        .onSet((newState) => this.setSpeed(newState));
    } else {
      this.service = new this.hap.Service.Switch(this.config.name, "Vacuum");
    }
    ensureName(this.hap, this.service, this.config.name);

    if (this.service.setPrimaryService) this.service.setPrimaryService(true);

    this.service
      .getCharacteristic(this.hap.Characteristic.On)
      .onGet(() => this.getCleaning())
      .onSet((newState) => this.setCleaning(newState))
      .on("change", ({ newValue }) => {
        this.changedPause(newValue === true);
      });
  }

  public async init() {
    if (this.config.serviceType === "fan") {
      this.deviceManager.deviceConnected$.subscribe(() => {
        // Now that we know the model, amend the steps in the Rotation speed (for better usability)
        const minStep =
          100 / (findSpeedModes(this.deviceManager.model).speed.length - 1);
        this.service
          .getCharacteristic(this.hap.Characteristic.RotationSpeed)
          .setProps({ minStep: minStep });
      });

      this.deviceManager.stateChanged$
        .pipe(
          filter(({ key }) => key === "fanSpeed"),
          distinct(({ value }) => value)
        )
        .subscribe(({ value: speed }) => {
          this.log.info(`MON changedSpeed | FanSpeed is now ${speed}%`);
          const speedMode = this.findSpeedModeFromMiio(speed);

          if (typeof speedMode === "undefined") {
            this.log.warn(
              `WAR changedSpeed | Speed was changed to ${speed}%, this speed is not supported`
            );
          } else {
            const { homekitTopLevel, name } = speedMode;
            this.log.info(
              `changedSpeed | Speed was changed to ${speed}% (${name}), for HomeKit ${homekitTopLevel}%`
            );
            this.service
              .getCharacteristic(this.hap.Characteristic.RotationSpeed)
              .updateValue(homekitTopLevel);
          }
        });
    }

    this.deviceManager.stateChanged$
      .pipe(
        filter(({ key }) => key === "cleaning"),
        map(({ value }) => value === true),
        tap((isCleaning) => {
          this.log.debug(
            `MON changedCleaning | CleaningState is now ${isCleaning}`
          );
          this.service
            .getCharacteristic(this.hap.Characteristic.On)
            .updateValue(isCleaning);
        }),
        distinct()
      )
      .subscribe((isCleaning) => {
        this.log.info(
          `changedCleaning | Cleaning is ${isCleaning ? "ON" : "OFF"}.`
        );
      });
  }

  get services() {
    return [this.service];
  }

  public async getCleaning() {
    try {
      const isCleaning = this.deviceManager.isCleaning;
      this.log.info(`getCleaning | Cleaning is ${isCleaning}`);

      return isCleaning;
    } catch (err) {
      this.log.error(`getCleaning | Failed getting the cleaning status.`, err);
      throw err;
    }
  }

  public async setCleaning(state) {
    await this.deviceManager.ensureDevice("setCleaning");
    this.log.info(`ACT setCleaning | Set cleaning to ${state}}`);
    try {
      if (state && !this.deviceManager.isCleaning) {
        // Start cleaning
        const roomIdsToClean = this.roomsService.roomIdsToClean;
        if (roomIdsToClean.size > 0) {
          await this.deviceManager.device.cleanRooms(
            Array.from(roomIdsToClean)
          );
          this.log.info(
            `ACT setCleaning | Start rooms cleaning for rooms ${Array.from(
              roomIdsToClean
            )}, device is in state ${this.deviceManager.property("state")}.`
          );
        } else {
          await this.deviceManager.device.activateCleaning();
          this.log.info(
            `ACT setCleaning | Start full cleaning, device is in state ${this.deviceManager.property(
              "state"
            )}.`
          );
        }
      } else if (
        !state &&
        (this.deviceManager.isCleaning || this.deviceManager.isPaused)
      ) {
        // Stop cleaning
        this.log.info(
          `ACT setCleaning | Stop cleaning and go to charge, device is in state ${this.deviceManager.property(
            "state"
          )}`
        );
        await this.deviceManager.device.activateCharging();
      }
    } catch (err) {
      this.log.error(`setCleaning | Failed to set cleaning to ${state}`, err);
      throw err;
    }
  }

  private async getSpeed() {
    await this.deviceManager.ensureDevice("getSpeed");

    const speed = await this.deviceManager.device.fanSpeed();
    this.log.info(
      `getSpeed | Fanspeed is ${speed} over miIO. Converting to HomeKit`
    );

    const { homekitTopLevel, name } = this.findSpeedModeFromMiio(speed) || {};

    this.log.info(
      `getSpeed | FanSpeed is ${speed} over miIO "${name}" > HomeKit speed ${homekitTopLevel}%`
    );

    return homekitTopLevel || 0;
  }

  public async setSpeed(speed) {
    await this.deviceManager.ensureDevice("setSpeed");

    if (typeof speed === "number") {
      this.log.debug(
        `ACT setSpeed | Speed got ${speed}% over HomeKit > CLEANUP.`
      );
    }

    // Get the speed modes for this model
    const speedModes = findSpeedModes(
      this.deviceManager.model,
      this.productInfo.firmware
    ).speed;

    let miLevel: number | null = null;
    let name: string | null = null;

    if (typeof speed === "number") {
      // Speed set by number
      // gen1 has maximum of 91%, so anything over that won't work. Getting safety maximum.
      const safeSpeed = Math.min(
        speed,
        speedModes[speedModes.length - 1].homekitTopLevel
      );

      // Find the minimum homekitTopLevel that matches the desired speed
      const speedMode = speedModes.find(
        (mode) => safeSpeed <= mode.homekitTopLevel
      )!;
      miLevel = speedMode.miLevel;
      name = speedMode.name;
    } else {
      // Set by mode name
      const speedMode = speedModes.find((mode) => mode.name === speed);

      if (speedMode == null) {
        this.log.info(`setSpeed | Mode "${speed}" does not exist.`);
        return;
      }
      miLevel = speedMode.miLevel;
      name = speedMode.name;
    }

    this.log.info(
      `ACT setSpeed | FanSpeed set to ${miLevel} over miIO for "${name}".`
    );

    // Save the latest set speed for handling the "custom" speed later
    this.cachedState.set("FanSpeed", miLevel);
    this.cachedState.set("FanSpeedName", name);

    if (miLevel === -1) {
      this.log.info(
        `setSpeed | FanSpeed is -1 => Calling setCleaning(false) instead of changing the fan speed`
      );
      await this.setCleaning(false);
    } else {
      await this.deviceManager.device.changeFanSpeed(miLevel);

      // If speed is "custom", also set the water speed to "custom" (for Xiaomi App)
      if (
        name === "Custom" &&
        this.config.waterBox &&
        this.cachedState.get("WaterSpeedName") !== "Custom"
      ) {
        await this.setWaterSpeed("Custom");
      }
      // If speed is not "custom" remove set the water speed also to a fixed value (for Xiaomi App)
      else if (
        name !== "Custom" &&
        this.config.waterBox &&
        this.cachedState.get("WaterSpeedName") === "Custom"
      ) {
        await this.setWaterSpeed("Medium");
      }
    }
  }

  private findSpeedModeFromMiio(speed) {
    // Get the speed modes for this model
    const speedModes = findSpeedModes(
      this.deviceManager.model,
      this.productInfo.firmware
    ).speed;

    // Find speed mode that matches the miLevel
    return speedModes.find((mode) => mode.miLevel === speed);
  }
}
