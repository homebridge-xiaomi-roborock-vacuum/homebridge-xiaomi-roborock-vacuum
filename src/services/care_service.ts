import { HAP, Service } from "homebridge";
import { CoreContext } from "./types";
import { callbackify } from "../utils/callbackify";
import { Care, Characteristic } from "./custom_care_service";
import { MainService } from "./main_service";
import { PluginServiceClass } from "./plugin_service_class";

export interface CareConfig {
  disableCareServices?: boolean;
  legacyCareSensors?: boolean;
}

export class CareService extends PluginServiceClass {
  public readonly services: Service[];
  constructor(
    coreContext: CoreContext,
    private readonly mainService: MainService
  ) {
    super(coreContext);
    this.services = this.config.legacyCareSensors
      ? this.registerLegacyCustomCareService()
      : this.registerNativeCareServices();
  }

  public async init(): Promise<void> {}

  private async getCareSensors() {
    // 30h = sensor_dirty_time
    const lifetime = 108000;
    const sensorDirtyTime =
      this.deviceManager.property<number>("sensorDirtyTime")!;
    const lifetimePercent = (sensorDirtyTime / lifetime) * 100;
    this.log.info(
      `getCareSensors | Sensors dirtyTime is ${sensorDirtyTime} seconds / ${lifetimePercent.toFixed(
        2
      )}%.`
    );
    return Math.min(100, lifetimePercent);
  }

  private async getCareFilter() {
    // 150h = filter_work_time
    const lifetime = 540000;
    const filterWorkTime =
      this.deviceManager.property<number>("filterWorkTime")!;
    const lifetimePercent = (filterWorkTime / lifetime) * 100;
    this.log.info(
      `getCareFilter | Filter workTime is ${filterWorkTime} seconds / ${lifetimePercent.toFixed(
        2
      )}%.`
    );
    return Math.min(100, lifetimePercent);
  }

  private async getCareSideBrush() {
    // 200h = side_brush_work_time
    const lifetime = 720000;
    const sideBrushWorkTime =
      this.deviceManager.property<number>("sideBrushWorkTime")!;
    const lifetimePercent = (sideBrushWorkTime / lifetime) * 100;
    this.log.info(
      `getCareSideBrush | SideBrush workTime is ${sideBrushWorkTime} seconds / ${lifetimePercent.toFixed(
        2
      )}%.`
    );
    return Math.min(100, lifetimePercent);
  }

  private async getCareMainBrush() {
    // 300h = main_brush_work_time
    const lifetime = 1080000;
    const mainBrushWorkTime =
      this.deviceManager.property<number>("mainBrushWorkTime")!;
    const lifetimePercent = (mainBrushWorkTime / lifetime) * 100;
    this.log.info(
      `getCareMainBrush | MainBrush workTime is ${mainBrushWorkTime} seconds / ${lifetimePercent.toFixed(
        2
      )}%.`
    );
    return Math.min(100, lifetimePercent);
  }

  private registerLegacyCustomCareService(): Service[] {
    const service = new Care(`${this.config.name} Care`);
    service
      .getCharacteristic(Characteristic.CareSensors)
      .on("get", (cb) => callbackify(() => this.getCareSensors(), cb));
    service
      .getCharacteristic(Characteristic.CareFilter)
      .on("get", (cb) => callbackify(() => this.getCareFilter(), cb));
    service
      .getCharacteristic(Characteristic.CareSideBrush)
      .on("get", (cb) => callbackify(() => this.getCareSideBrush(), cb));
    service
      .getCharacteristic(Characteristic.CareMainBrush)
      .on("get", (cb) => callbackify(() => this.getCareMainBrush(), cb));

    return [service];
  }

  private registerNativeCareServices(): Service[] {
    const mainService = this.mainService.services[0]; // Hack for now

    // Register the high-level state of the filters to the main device's Characteristics
    mainService
      .getCharacteristic(this.hap.Characteristic.FilterChangeIndication)
      .on("get", (cb) =>
        callbackify(async () => {
          const carePercentages = await Promise.all([
            this.getCareSensors(),
            this.getCareFilter(),
            this.getCareSideBrush(),
            this.getCareMainBrush(),
          ]);
          return carePercentages.some((item) => item >= 100);
        }, cb)
      );
    mainService
      .getCharacteristic(this.hap.Characteristic.FilterLifeLevel)
      .on("get", (cb) =>
        callbackify(async () => {
          const carePercentages = await Promise.all([
            this.getCareSensors(),
            this.getCareFilter(),
            this.getCareSideBrush(),
            this.getCareMainBrush(),
          ]);
          return 100 - Math.max(...carePercentages);
        }, cb)
      );

    // Use Homekit's native FilterMaintenance Service
    const careSensorsService = new this.hap.Service.FilterMaintenance(
      "Care indicator sensors",
      "sensors"
    );
    careSensorsService
      .getCharacteristic(this.hap.Characteristic.FilterChangeIndication)
      .on("get", (cb) =>
        callbackify(async () => {
          return (await this.getCareSensors()) >= 100;
        }, cb)
      );
    careSensorsService
      .getCharacteristic(this.hap.Characteristic.FilterLifeLevel)
      .on("get", (cb) =>
        callbackify(async () => 100 - (await this.getCareSensors()), cb)
      );

    const careFilterService = new this.hap.Service.FilterMaintenance(
      "Care indicator filter",
      "filter"
    );
    careFilterService
      .getCharacteristic(this.hap.Characteristic.FilterChangeIndication)
      .on("get", (cb) =>
        callbackify(async () => {
          return (await this.getCareFilter()) >= 100;
        }, cb)
      );
    careFilterService
      .getCharacteristic(this.hap.Characteristic.FilterLifeLevel)
      .on("get", (cb) =>
        callbackify(async () => 100 - (await this.getCareFilter()), cb)
      );

    const careSideBrushService = new this.hap.Service.FilterMaintenance(
      "Care indicator side brush",
      "side brush"
    );
    careSideBrushService
      .getCharacteristic(this.hap.Characteristic.FilterChangeIndication)
      .on("get", (cb) =>
        callbackify(async () => {
          return (await this.getCareSideBrush()) >= 100;
        }, cb)
      );
    careSideBrushService
      .getCharacteristic(this.hap.Characteristic.FilterLifeLevel)
      .on("get", (cb) =>
        callbackify(async () => 100 - (await this.getCareSideBrush()), cb)
      );

    const careMainBrushService = new this.hap.Service.FilterMaintenance(
      "Care indicator main brush",
      "main brush"
    );
    careMainBrushService
      .getCharacteristic(this.hap.Characteristic.FilterChangeIndication)
      .on("get", (cb) =>
        callbackify(async () => {
          return (await this.getCareMainBrush()) >= 100;
        }, cb)
      );
    careMainBrushService
      .getCharacteristic(this.hap.Characteristic.FilterLifeLevel)
      .on("get", (cb) =>
        callbackify(async () => 100 - (await this.getCareMainBrush()), cb)
      );

    return [
      careSensorsService,
      careFilterService,
      careSideBrushService,
      careMainBrushService,
    ];
  }
}
