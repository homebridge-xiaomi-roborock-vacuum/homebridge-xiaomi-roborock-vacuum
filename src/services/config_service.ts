import { CustomLoggerConfig } from "../utils/logger";
import { RoomsConfig } from "./rooms_service";
import { DeviceManagerConfig } from "./device_manager";
import { PauseConfig } from "./pause_switch";
import { WaterBoxConfig } from "./water_box_service";
import { DustCollectionConfig } from "./dust_collection";
import { FindMeConfig } from "./find_me_service";
import { GoToConfig } from "./go_to_service";
import { DockConfig } from "./dock_service";
import { ZonesConfig } from "./zones_service";
import { CareConfig } from "./care_service";

export interface Config
  extends DeviceManagerConfig,
    CustomLoggerConfig,
    RoomsConfig,
    PauseConfig,
    WaterBoxConfig,
    DustCollectionConfig,
    FindMeConfig,
    GoToConfig,
    DockConfig,
    ZonesConfig,
    CareConfig {
  /**
   * The name of the main service as it will show up in the Home App.
   */
  name: string;

  // For now, let's allow anything until all config entries are defined
  [key: string]: any;
}

/**
 * Applies the default configuration values to the config provided by the user.
 * @param config The config provided by the user.
 *
 * @remarks We may want to use a validation library like io-ts in the future for easier typing and defaulting.
 */
export function applyConfigDefaults(config: Partial<Config>): Config {
  return {
    name: "Roborock vacuum cleaner",
    cleanword: "cleaning",
    pause: false,
    pauseWord: "Pause",
    findMe: false,
    findMeWord: "where are you",
    goTo: false,
    goToWord: "go to coordinates",
    goToX: 25500,
    goToY: 25500,
    roomTimeout: 0,
    waterBox: false,
    dustCollection: false,
    dock: false,
    ...config,
  };
}
