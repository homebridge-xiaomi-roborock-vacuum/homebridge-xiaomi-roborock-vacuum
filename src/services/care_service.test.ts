import { CareService } from "./care_service";
import { createHomebridgeMock } from "../test.mocks";
import { getLoggerMock } from "../utils/logger.mock";
import { applyConfigDefaults } from "./config_service";
import {
  createDeviceManagerMock,
  DeviceManagerMock,
} from "./device_manager.mock";
import { MainService } from "./main_service";
import { DeviceManager } from "./device_manager";
import { HAP } from "homebridge";
import * as HapJs from "hap-nodejs";

describe("CareService", () => {
  describe("legacyCareSensors", () => {
    let careService: CareService;
    let deviceManagerMock: DeviceManagerMock;
    let fanServices: jest.Mock;
    let hap: jest.Mocked<HAP>;

    beforeEach(() => {
      const hbMock = createHomebridgeMock();
      hap = hbMock.hap as any;
      const log = getLoggerMock();

      deviceManagerMock = createDeviceManagerMock();
      deviceManagerMock.property.mockReturnValue(10);

      fanServices = jest.fn();
      const fan = {
        get services() {
          return fanServices();
        },
      } as MainService;

      const config = applyConfigDefaults({ legacyCareSensors: true });

      careService = new CareService(
        {
          hap,
          log,
          config,
          deviceManager: deviceManagerMock as unknown as DeviceManager,
        },
        fan
      );
    });

    test("fan is not called", () => {
      expect(fanServices).not.toHaveBeenCalled();
    });

    test("only 1 service is exposed", () => {
      expect(careService.services).toHaveLength(1);
    });

    test("the custom Care service characteristics call all getters", async () => {
      const service = careService.services[0];
      expect(service.characteristics).toHaveLength(5);
      const [name, ...characteristics] = service.characteristics;
      const onCallsHandlers = characteristics.map(
        (characteristic) => characteristic["_events"].get
      );
      for (const listener of onCallsHandlers) {
        await expect(
          new Promise((resolve, reject) =>
            listener((err, value) => (err ? reject(err) : resolve(value)))
          )
        ).resolves.toMatchSnapshot();
      }
    });
  });

  describe("nativeCareSensors", () => {
    let careService: CareService;
    let deviceManagerMock: DeviceManagerMock;
    let fanServices: jest.Mock;
    let hap: HAP;

    beforeEach(() => {
      const hbMock = createHomebridgeMock();
      hap = HapJs;
      const log = getLoggerMock();

      deviceManagerMock = createDeviceManagerMock();
      deviceManagerMock.property.mockReturnValue(10);

      fanServices = jest
        .fn()
        .mockImplementation(() => [new hbMock.hap.Service.Fan()]);
      const fan = {
        get services() {
          return fanServices();
        },
      } as MainService;

      const config = applyConfigDefaults({});

      careService = new CareService(
        {
          hap,
          log,
          config,
          deviceManager: deviceManagerMock as unknown as DeviceManager,
        },
        fan
      );
    });

    test("fan is called", () => {
      expect(fanServices).toHaveBeenCalledTimes(1);
    });

    test("4 services are exposed", () => {
      expect(careService.services).toHaveLength(4);
    });

    test("the fan's service characteristic calls all getters", async () => {
      const fanService = fanServices.mock.results[0].value[0];
      expect(fanService.getCharacteristic.mock.results).toHaveLength(2);
      const onCallsHandlers = fanService.getCharacteristic.mock.results.map(
        ({ value }) => value.on.mock.calls[0][1]
      );
      await expect(
        new Promise((resolve, reject) =>
          onCallsHandlers[0]((err, value) =>
            err ? reject(err) : resolve(value)
          )
        )
      ).resolves.toBe(false);
      await expect(
        new Promise((resolve, reject) =>
          onCallsHandlers[1]((err, value) =>
            err ? reject(err) : resolve(value)
          )
        )
      ).resolves.toBe(99.99074074074075);
    });

    test("the FilterMaintenance services' characteristics call all getters", async () => {
      for (const service of careService.services) {
        expect(service.characteristics).toHaveLength(3);
        const [name, ...characteristics] = service.characteristics;
        const onCallsHandlers = characteristics.map(
          (characteristic) => characteristic["_events"].get
        );
        for (const listener of onCallsHandlers) {
          await expect(
            new Promise((resolve, reject) =>
              listener((err, value) => (err ? reject(err) : resolve(value)))
            )
          ).resolves.toMatchSnapshot();
        }
      }
    });

    test("handles -1 responses", async () => {
      for (const service of careService.services) {
        expect(service.characteristics).toHaveLength(3);
        const [name, ...characteristics] = service.characteristics;
        const onCallsHandlers = characteristics.map(
          (characteristic) => characteristic["_events"].get
        );
        for (const listener of onCallsHandlers) {
          deviceManagerMock.property.mockReturnValueOnce(-1);
          await expect(
            new Promise((resolve, reject) =>
              listener((err, value) => (err ? reject(err) : resolve(value)))
            )
          ).resolves.toMatchSnapshot();
        }
      }
    });
  });
});
