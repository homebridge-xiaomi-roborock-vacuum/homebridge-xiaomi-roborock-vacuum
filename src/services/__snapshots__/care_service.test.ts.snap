// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CareService legacyCareSensors the custom Care service characteristics call all getters 1`] = `0.009259259259259259`;

exports[`CareService legacyCareSensors the custom Care service characteristics call all getters 2`] = `0.001851851851851852`;

exports[`CareService legacyCareSensors the custom Care service characteristics call all getters 3`] = `0.001388888888888889`;

exports[`CareService legacyCareSensors the custom Care service characteristics call all getters 4`] = `0.000925925925925926`;

exports[`CareService nativeCareSensors the FilterMaintenance services' characteristics call all getters 1`] = `false`;

exports[`CareService nativeCareSensors the FilterMaintenance services' characteristics call all getters 2`] = `99.99074074074075`;

exports[`CareService nativeCareSensors the FilterMaintenance services' characteristics call all getters 3`] = `false`;

exports[`CareService nativeCareSensors the FilterMaintenance services' characteristics call all getters 4`] = `99.99814814814815`;

exports[`CareService nativeCareSensors the FilterMaintenance services' characteristics call all getters 5`] = `false`;

exports[`CareService nativeCareSensors the FilterMaintenance services' characteristics call all getters 6`] = `99.99861111111112`;

exports[`CareService nativeCareSensors the FilterMaintenance services' characteristics call all getters 7`] = `false`;

exports[`CareService nativeCareSensors the FilterMaintenance services' characteristics call all getters 8`] = `99.99907407407407`;
